packages/agreement-process/src/model/domain/validators.ts

```diff
@@ -0,0 +1,341 @@
+
+  if (agreements.length > 0) {
+    throw agreementAlreadyExists(consumerId, eserviceId);
+  }
+};
+
+export function assertAgreementExist(
+  agreementId: string,
+  agreement: WithMetadata<Agreement> | undefined
+): asserts agreement is NonNullable<WithMetadata<Agreement>> {
+  if (agreement === undefined) {
+    throw agreementNotFound(agreementId);
+  }
+}
+
+export const assertRequesterIsConsumer = (
+  consumerId: string,
+  requesterId: string
+): void => {
+  if (consumerId !== requesterId) {
+    throw operationNotAllowed(requesterId);
+  }
+};
+
+export const validateCreationOnDescriptor = (
+  eservice: EService,
+  descriptorId: string
+): Descriptor => {
+  const allowedStatus = [descriptorState.published];
+  return validateLatestDescriptor(eservice, descriptorId, allowedStatus);
+};
+
+export const verifyCreationConflictingAgreements = async (
+  organizationId: string,
+  agreement: ApiAgreementPayload,
+  agreementQuery: AgreementQuery
+): Promise<void> => {
+  const conflictingStates: AgreementState[] = [
+    agreementState.draft,
+    agreementState.pending,
+    agreementState.missingCertifiedAttributes,
+    agreementState.active,
+    agreementState.suspended,
+  ];
+  await verifyConflictingAgreements(
+    organizationId,
+    agreement.eserviceId,
+    conflictingStates,
+    agreementQuery
+  );
+};
+
+export const validateCertifiedAttributes = (
+  descriptor: Descriptor,
+  consumer: Tenant
+): void => {
+  if (!certifiedAttributesSatisfied(descriptor, consumer)) {
+    throw missingCertifiedAttributesError(descriptor.id, consumer.id);
+  }
+};
+
+export const assertExpectedState = (
+  agreementId: string,
+  agreementState: AgreementState,
+  expectedStates: AgreementState[]
+): void => {
+  if (!expectedStates.includes(agreementState)) {
+    throw agreementNotInExpectedState(agreementId, agreementState);
+  }
+};
+export const certifiedAttributesSatisfied = (
+  descriptor: Descriptor,
+  tenant: Tenant
+): boolean => {
+  const certifiedAttributes = tenant.attributes.filter(
+    (e) => e.type === tenantAttributeType.CERTIFIED
+  ) as CertifiedTenantAttribute[];
+
+  return attributesSatisfied(
+    descriptor.attributes.certified,
+    certifiedAttributes
+  );
+};
+
+export const declaredAttributesSatisfied = (
+  descriptor: Descriptor,
+  tenant: Tenant
+): boolean => {
+  const declaredAttributes = tenant.attributes.filter(
+    (e) => e.type === tenantAttributeType.DECLARED
+  ) as DeclaredTenantAttribute[];
+
+  return attributesSatisfied(
+    descriptor.attributes.declared,
+    declaredAttributes
+  );
+};
+
+export const verifiedAttributesSatisfied = (
+  agreement: Agreement,
+  descriptor: Descriptor,
+  tenant: Tenant
+): boolean => {
+  const verifiedAttributes = tenant.attributes.filter(
+    (e) => e.type === tenantAttributeType.DECLARED
+  ) as VerifiedTenantAttribute[];
+
+  const producersAttributesNotExpired = verifiedAttributes.filter(
+    (a): boolean =>
+      !!a.verifiedBy.find(
```

Comment:
I may be missing something, but shouldn't this be
```suggestion
      a.verifiedBy.find(
```
?

Reply:
it should be a invalid typecheck error, in this case the attribute `a` can never be `undefined`, from this prospective the `!!` operator seems an acceptable choice, anyway I rewrite that with more explicit check.

Reply:
it should be a invalid typecheck error, in this case the attribute `a` can never be `undefined`, from this prospective the `!!` operator seems an acceptable choice, anyway I rewrite that with more explicit check.

Reply:
oh ok, then it's fine! my bad 😄